packages/core/src/node/messaging/ipc-connection-provider.ts

```diff
@@ -74,32 +76,41 @@ export class IPCConnectionProvider {
         return toStop;
     }
 
-    protected createConnection(childProcess: cp.ChildProcess, options: ResolvedIPCConnectionOptions): MessageConnection {
-        const reader = new IPCMessageReader(childProcess);
-        const writer = new IPCMessageWriter(childProcess);
-        const connection = createMessageConnection(reader, writer, {
-            error: (message: string) => this.logger.error(`[${options.serverName}: ${childProcess.pid}] ${message}`),
-            warn: (message: string) => this.logger.warn(`[${options.serverName}: ${childProcess.pid}] ${message}`),
-            info: (message: string) => this.logger.info(`[${options.serverName}: ${childProcess.pid}] ${message}`),
-            log: (message: string) => this.logger.info(`[${options.serverName}: ${childProcess.pid}] ${message}`)
+    protected createConnection(childProcess: cp.ChildProcess, options?: ResolvedIPCConnectionOptions): Channel {
+
+        const onCloseEmitter = new Emitter<ChannelCloseEvent>();
+        const onMessageEmitter = new Emitter<MessageProvider>();
+        const onErrorEmitter = new Emitter<unknown>();
+        const pipe = childProcess.stdio[4] as Writable;
+
+        pipe.on('data', (data: Uint8Array) => {
+            onMessageEmitter.fire(() => new Uint8ArrayReadBuffer(data));
         });
```

Comment:
Not sure I understand correctly but I believe there's the potential for a bug:

`stdio[4]` seems to assume that we are dealing with a forked js script, so there's an extra pipe opened for IPC.

The problem I see here is that

1. It feels weird to use this fd which is normally reserved for Node's `process.send` API.
2. Using the fd directly means we're dealing with a buffered stream and not a "whole message" connection.

Does the RPC protocol handle such streams properly? As in what you send might get chopped/chunked arbitrarily?

Reply:
>It feels weird to use this fd which is normally reserved for Node's process.send API.

Actually we are using an extra fd/pipe for directly sending binary data.
stido[0]= stdin
stdio[1]= stdout
stdio[2]= stderr
stdtio[3] = IPC   (reserved for node Node's `process.send` API)
stdio[4]  = Our extra binary pipe

We have to use an extra pipe because the IPC channel cannot be used to send binary data. It only supports JSON objects and will always invoke `JSON.stringify` before sending the message.

>Using the fd directly means we're dealing with a buffered stream and not a "whole message" connection.
Does the RPC protocol handle such streams properly? As in what you send might get chopped/chunked arbitrarily?

Good catch. I believe at the moment this is not handled properly. However, we implemented a solution for this problem in https://github.com/eclipse-theia/theia/pull/11093, that can handle binary messages as part of a buffered stream. I will update the IPC connection provider to also make use of this solution.